### PR TITLE
Refactor ModelBuilder tests and add more coverage.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Remove redundant CollectionReferenceBuilder

Part of #1102

### CODE REVIEW

File: src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs [hunk 1]
```diff
@@ -80,11 +80,16 @@ public virtual EntityTypeBuilder Annotation([NotNull] string annotation, [NotNul
...
         public virtual KeyBuilder Key([NotNull] params string[] propertyNames)
         {
-            Check.NotNull(propertyNames, nameof(propertyNames));
+            Check.NotEmpty(propertyNames, nameof(propertyNames));
 
             return new KeyBuilder(Builder.PrimaryKey(propertyNames, ConfigurationSource.Explicit));
         }
 
+        /// <summary>
+        ///     Creates a new unique index for this entity type if one does not already exist over the specified properties.
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs]: Nope! This is specifcially NOT about creating a unique "index". :-) This is for cases where you want an unique constraint on a column, but don't want an index, and don't want it to be used at one end or the other of a relationship. (Or, put differently, you might want those things, in which case you will need to make some other call anyway and this call will be redundant.)

File: src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs [hunk 2]
```diff
@@ -80,11 +80,16 @@ public virtual EntityTypeBuilder Annotation([NotNull] string annotation, [NotNul
...
+        /// </summary>
+        /// <param name="propertyNames"> The names of the properties that make up the unique index. </param>
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs]: Not an index.

File: src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs [hunk 3]
```diff
@@ -80,11 +80,16 @@ public virtual EntityTypeBuilder Annotation([NotNull] string annotation, [NotNul
...
+        /// <returns> An object that can be used to configure the unique index. </returns>
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs]: Not an index.

File: src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder`.cs [hunk 4]
```diff
@@ -74,6 +74,19 @@ public virtual KeyBuilder Key([NotNull] Expression<Func<TEntity, object>> keyExp
             return new KeyBuilder(Builder.PrimaryKey(keyExpression.GetPropertyAccessList(), ConfigurationSource.Explicit));
         }
 
+        /// <summary>
+        ///     Creates a new unique index for this entity type if one does not already exist over the specified properties.
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder`.cs]: Not an index,

File: test/EntityFramework.Core.Tests/ModelBuilderGenericTest.cs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        {
+            public GenericTestEntityTypeBuilder(EntityTypeBuilder<TEntity> entityTypeBuilder)
+            {
+                EntityTypeBuilder = entityTypeBuilder;
+            }
+
+            private EntityTypeBuilder<TEntity> EntityTypeBuilder { get; }
+            public override EntityType Metadata => EntityTypeBuilder.Metadata;
+
+            public override TestEntityTypeBuilder<TEntity> Annotation(string annotation, object value)
```
Comment (ajcvickers) [inline test/EntityFramework.Core.Tests/ModelBuilderGenericTest.cs]: Consider expression body syntax for lots of these methods.

**Comment by ajcvickers** [comment]
:shipit:

File: src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs (hunk 1 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs]: We call it "unique index" in the .PrincipalKey docs. Should that be changed to "unique constraint" as well?

Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Builders/EntityTypeBuilder.cs]: Probably, yes.

**Pull request merged**